File: go/Rectangle.go
package main

import (
	"fmt"
	"math"
)

type Rectangle struct {
	start Point
	end   Point
}

//---------------------------------------------------------------------------------------
func (thisPt *Rectangle) SetStart(p *Point) {
	thisPt.start = *p
}

//---------------------------------------------------------------------------------------
func (thisPt *Rectangle) SetEnd(p *Point) {
	thisPt.end = *p
}

//---------------------------------------------------------------------------------------
func (thisPt *Rectangle) PointInside(point *Point) bool {
	checkV := func(start float64, end float64, val float64) bool {
		if val >= start && val <= end {
			return true
		}
		if val <= start && val >= end {
			return true
		}
		return false
	}
	return (checkV(thisPt.start.GetX(), thisPt.end.GetX(), point.GetX()) &&
		checkV(thisPt.start.GetY(), thisPt.end.GetY(), point.GetY()))
}

//---------------------------------------------------------------------------------------
func (thisPt *Rectangle) GetStart() *Point {
	return &thisPt.start
}

//---------------------------------------------------------------------------------------
func (thisPt *Rectangle) GetEnd() *Point {
	return &thisPt.end
}

//---------------------------------------------------------------------------------------
func (thisPt *Rectangle) GetLength() Point {
	p := Point{}
	p.SetX(thisPt.end.GetX() - thisPt.start.GetX())
	p.SetY(thisPt.end.GetY() - thisPt.start.GetY())
	return p
}

//---------------------------------------------------------------------------------------
func (thisPt *Rectangle) Add(other *Rectangle) {
	nStart := Point{}
	nEnd := Point{}
	nStart.SetX(math.Min(other.start.GetX(), thisPt.start.GetX()))
	nStart.SetY(math.Min(other.start.GetY(), thisPt.start.GetY()))
	nEnd.SetX(math.Max(other.end.GetX(), thisPt.end.GetX()))
	nEnd.SetY(math.Max(other.end.GetY(), thisPt.end.GetY()))
	thisPt.start = nStart
	thisPt.end = nEnd
}

//---------------------------------------------------------------------------------------
func (thisPt *Rectangle) GetWKTString() string {
	out := "POLYGON  (("
	out += fmt.Sprintf("%f %f,%f %f,%f %f,%f %f,%f %f",
		thisPt.start.GetX(), thisPt.start.GetY(),
		thisPt.end.GetX(), thisPt.start.GetY(),
		thisPt.end.GetX(), thisPt.end.GetY(),
		thisPt.start.GetX(), thisPt.end.GetY(),
		thisPt.start.GetX(), thisPt.start.GetY())

	out += "))"
	return out
}

File: rust/polygon_search/src/geo/rectangle.rs
use crate::geo::point::Point;

#[derive(Copy, Clone)]
pub struct Rect {
    start_: Point,
    end_: Point,
}
//---------------------------------------------------------------------------------------
impl Rect {
    pub fn new(start: Point, end: Point) -> Rect {
        return Rect {
            start_: start,
            end_: end,
        };
    }

    pub fn get_start(&self) -> &Point {
        return &self.start_;
    }

    pub fn get_end(&self) -> &Point {
        return &self.end_;
    }

    pub fn set_start(&mut self, p: &Point) {
        self.start_ = *p;
    }

    pub fn set_end(&mut self, p: &Point) {
        self.end_ = *p;
    }

    pub fn point_inside(&self, p: &Point) -> bool {
        let check_v = |start: f64, end: f64, val: f64| -> bool {
            if val >= start  && val <= end{
                return true;
            }
            if val <= start && val >= end {
                return true;
            }
            return false;
        };

        return check_v(self.start_.get_x(), self.end_.get_x(), p.get_x())
            && check_v(self.start_.get_y(), self.end_.get_y(), p.get_y());
    }

    
    pub fn add(&mut self, other: &Rect) {

        let n_start = Point::new(
            other.get_start().get_x().min(self.get_start().get_x()),
            other.get_start().get_y().min(self.get_start().get_y()),
        );

        let n_end = Point::new(
            other.get_end().get_x().max(self.get_end().get_x()),
            other.get_end().get_y().max(self.get_end().get_y()),
        );

        self.set_start(&n_start);
        self.set_end(&n_end);
    }

    pub fn get_length(&self) -> Point {
        let l=self.end_.get_x()-self.start_.get_x();
        let h=self.end_.get_y()-self.start_.get_y();
        return Point::new(l,h);
    }
}
